chia/wallet/wallet_protocol.py

```diff
@@ -0,0 +1,65 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, List, Optional, Set
+
+from blspy import G1Element
+from typing_extensions import Protocol
+
+from chia.server.ws_connection import WSChiaConnection
+from chia.types.blockchain_format.coin import Coin
+from chia.types.blockchain_format.sized_bytes import bytes32
+from chia.util.ints import uint8, uint32, uint64, uint128
+from chia.wallet.wallet_coin_record import WalletCoinRecord
+
+if TYPE_CHECKING:
+    from chia.wallet.wallet_state_manager import WalletStateManager
+
+
+class WalletProtocol(Protocol):
+
+    # TODO: it seems like this should return WalletType instead
+    @classmethod
+    def type(cls) -> uint8:
+        ...
+
+    def id(self) -> uint32:
+        ...
+
+    async def coin_added(self, coin: Coin, height: uint32, peer: WSChiaConnection) -> None:
+        ...
+
+    async def select_coins(
+        self,
+        amount: uint64,
+        exclude: Optional[List[Coin]] = None,
+        min_coin_amount: Optional[uint64] = None,
+        max_coin_amount: Optional[uint64] = None,
+    ) -> Set[Coin]:
```

Comment:
This function had a discrepancy with the `DIDWallet`, which used to return an `Optional` instead of throwing on errors. Most wallets used this signature, so I harmonised them to this one.

see: https://github.com/Chia-Network/chia-blockchain/pull/13437/files#diff-ff2dff17865341c5c8414c162e4d4031faf867bece63a892edf43ff4629d2631L309